libraries/ms-common/src/can_datagram.c

```diff
@@ -0,0 +1,634 @@
+  FSM_ADD_TRANSITION(store->rx_event, state_rx_crc);
+  FSM_ADD_TRANSITION(store->repeat_event, state_rx_protocol_version);
+  FSM_ADD_TRANSITION(store->error_event, state_done);
+}
+
+FSM_STATE_TRANSITION(state_rx_crc) {
+  CanDatagramStorage *store = fsm->context;
+  FSM_ADD_TRANSITION(store->rx_event, state_rx_dgram_type);
+  FSM_ADD_TRANSITION(store->repeat_event, state_rx_crc);
+  FSM_ADD_TRANSITION(store->error_event, state_done);
+}
+
+FSM_STATE_TRANSITION(state_rx_dgram_type) {
+  CanDatagramStorage *store = fsm->context;
+  FSM_ADD_TRANSITION(store->rx_event, state_rx_dst_len);
+  FSM_ADD_TRANSITION(store->repeat_event, state_rx_dgram_type);
+  FSM_ADD_TRANSITION(store->error_event, state_done);
+}
+
+FSM_STATE_TRANSITION(state_rx_dst_len) {
+  CanDatagramStorage *store = fsm->context;
+  FSM_ADD_TRANSITION(store->rx_event, state_rx_dst);
+  FSM_ADD_TRANSITION(store->repeat_event, state_rx_dst_len);
+  FSM_ADD_TRANSITION(store->error_event, state_done);
+}
+
+FSM_STATE_TRANSITION(state_rx_dst) {
+  CanDatagramStorage *store = fsm->context;
+  FSM_ADD_TRANSITION(store->rx_event, state_rx_data_len);
+  FSM_ADD_TRANSITION(store->repeat_event, state_rx_dst);
+  FSM_ADD_TRANSITION(store->error_event, state_done);
+}
+
+FSM_STATE_TRANSITION(state_rx_data_len) {
+  CanDatagramStorage *store = fsm->context;
+  FSM_ADD_TRANSITION(store->rx_event, state_rx_data);
+  FSM_ADD_TRANSITION(store->repeat_event, state_rx_data_len);
+  FSM_ADD_TRANSITION(store->error_event, state_done);
+}
+
+FSM_STATE_TRANSITION(state_rx_data) {
+  CanDatagramStorage *store = fsm->context;
+  FSM_ADD_TRANSITION(store->rx_event, state_done);
+  FSM_ADD_TRANSITION(store->repeat_event, state_rx_data);
+  FSM_ADD_TRANSITION(store->error_event, state_done);
+}
+
+FSM_STATE_TRANSITION(state_done) {
+  CanDatagramStorage *store = fsm->context;
+  FSM_ADD_TRANSITION(store->rx_event, state_idle);
+  FSM_ADD_TRANSITION(store->tx_event, state_idle);
+  FSM_ADD_TRANSITION(store->error_event, state_idle);
+}
+
+//  Write max 8 bytes to fifo on datagram frame rx
+static void prv_write_rx_buffer(uint8_t *data, size_t len) {
+  int ret = fifo_push_arr(&s_fifo, data, len);
+  // Catch buffer overflow as error
+  if (ret == STATUS_CODE_RESOURCE_EXHAUSTED) {
+    LOG_WARN("RX MSG BUFFER OVERFLOW. EXITING WITH ERROR...\n");
+    event_raise(s_store.error_event, HARD_ERROR);
+  }
+}
+
+// Read array of max length 2048 direct from fifo
+static bool prv_read_rx_buffer(uint8_t *data, size_t len) {
+  // Check that data exists to be read
+  return (fifo_pop_arr(&s_fifo, data, len) != STATUS_CODE_RESOURCE_EXHAUSTED);
+}
+
+// Functions used to write and read to fifo
+static void prv_flush_tx_buffer(void) {
+  size_t tx_len = fifo_size(&s_fifo) < CAN_BUFFER_SIZE ? fifo_size(&s_fifo) : CAN_BUFFER_SIZE;
+  for (size_t i = 0; i < tx_len; i++) {
+    fifo_pop(&s_fifo, &s_can_buffer[i]);
+  }
+  fifo_pop_arr(&s_fifo, s_can_buffer, tx_len);
+  s_store.tx_cb(s_can_buffer, tx_len, false);
```

Comment:
Even though you've already null checked elsewhere, maybe surround this with a null check just in case :)